### PR TITLE
feat: update DPP to 0.16.0

### PR DESCRIPTION
**Pull request by shuplenkov**
## Issue being fixed or feature implemented

Update to new protocol version 0.16

## What was done?

- DPP updated to 0.16.0-dev
- Query handlers accept and respond with Buffers
- Identifier is used instead of strings
- Removed deprecated `/identity/by-first-public-key` and `/identityIds/by-first-public-key` queries
- Store keys as Buffers in leveldb

## How Has This Been Tested?

With tests

## Breaking Changes

- `getIdentityByFirstPublicKey` method has been removed
- `getIdentityIdByFirstPublicKey` method has been removed
- `fetchPublicKeyIdentityId` method has been removed
- `storePublicKeyIdentityId` method has been removed
- `where` clause in document query now awaits binary (`Buffer` compatible) data for `ids`
- `/identities` query now awaits `id` as a Buffer in data
- `/data-contracts/` query now awaits `id` as a Buffer in data

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have made corresponding changes to the documentation

**For repository code-owners and collaborators only**
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

File: lib/dataContract/DataContractLevelDBRepository.js [hunk 1]
```diff
@@ -22,8 +23,8 @@ class DataContractLevelDBRepository {
     const db = transaction ? transaction.db : this.db;
 
     await db.put(
-      this.addKeyPrefix(dataContract.getId()),
-      dataContract.serialize(),
+      this.addKeyPrefix(bs58.encode(dataContract.getId())),
```
Comment (shumkov) [inline lib/dataContract/DataContractLevelDBRepository.js:26]: dataContract.getId().toString()

File: lib/document/mongoDbRepository/DocumentMongoDbRepository.js [hunk 2]
```diff
@@ -288,10 +288,10 @@ class DocumentMongoDbRepository {
     protocolVersion,
     data,
   }) {
-    const encodedProperties = this.dataContract.getEncodedProperties(this.documentType);
+    const binaryProperties = this.dataContract.getBinaryProperties(this.documentType);
 
-    Object.entries(encodedProperties).forEach(([propertyPath, propertyValue]) => {
-      if (propertyValue.contentEncoding === 'base64') {
+    Object.entries(binaryProperties).forEach(([propertyPath, propertyValue]) => {
+      if (propertyValue.contentMediaType === 'application/x.dash.dpp.identifier') {
```
Comment (shumkov) [inline lib/document/mongoDbRepository/DocumentMongoDbRepository.js:294]: You need to convert all binary properties from MongoDB Blob to Buffer

File: lib/document/mongoDbRepository/DocumentMongoDbRepository.js [hunk 3]
```diff
@@ -288,10 +288,10 @@ class DocumentMongoDbRepository {
     protocolVersion,
     data,
   }) {
-    const encodedProperties = this.dataContract.getEncodedProperties(this.documentType);
+    const binaryProperties = this.dataContract.getBinaryProperties(this.documentType);
```
Comment (shumkov) [inline lib/document/mongoDbRepository/DocumentMongoDbRepository.js:291]: You can pass `promoteBuffers: true` option to `find` methods to get back native Buffers so you won't need to do this conversion.

File: lib/identity/IdentityLevelDBRepository.js [hunk 4]
```diff
@@ -22,8 +23,8 @@ class IdentityLevelDBRepository {
     const db = transaction ? transaction.db : this.db;
 
     await db.put(
-      this.addKeyPrefix(identity.getId()),
-      identity.serialize(),
+      this.addKeyPrefix(bs58.encode(identity.getId())),
```
Comment (shumkov) [inline lib/identity/IdentityLevelDBRepository.js:26]: identity.getId().toString()

File: test/integration/dataContract/DataContractLevelDBRepository.spec.js [hunk 5]
```diff
@@ -24,12 +25,12 @@ describe('DataContractLevelDBRepository', () => {
...
     dppMock
       .dataContract
-      .createFromSerialized
+      .createFromBuffer
       .resolves(dataContract);
 
     repository = new DataContractLevelDBRepository(db, dppMock);
 
-    key = `${DataContractLevelDBRepository.KEY_PREFIX}:${dataContract.getId()}`;
+    key = `${DataContractLevelDBRepository.KEY_PREFIX}:${bs58.encode(dataContract.getId())}`;
```
Comment (shumkov) [inline test/integration/dataContract/DataContractLevelDBRepository.spec.js:33]: You can put just `dataContract.getId()` and `toString()` will be called automatically.

**Review by shumkov — DISMISSED** [review]
👍

File: lib/abci/handlers/query/documentQueryHandlerFactory.js [hunk 6]
```diff
@@ -18,7 +18,7 @@ function documentQueryHandlerFactory(fetchDocuments) {
   /**
    * @typedef documentQueryHandler
    * @param {Object} params
-   * @param {string} params.contractId
+   * @param {Buffer} params.contractId
```
Comment (shumkov) [inline lib/abci/handlers/query/documentQueryHandlerFactory.js:21]: It should be in data (called options here)

File: lib/createDIContainer.js [hunk 7]
```diff
@@ -411,25 +405,19 @@ async function createDIContainer(options) {
...
       identityIdsByPublicKeyHashesQueryHandler,
     ) => {
       const router = findMyWay({
         ignoreTrailingSlash: true,
       });
 
-      router.on('GET', '/identities/:id', identityQueryHandler);
-      router.on('GET', '/dataContracts/:id', dataContractQueryHandler);
-      router.on('GET', '/dataContracts/:contractId/documents/:type', documentQueryHandler);
+      router.on('GET', '/identities/by-id', identityQueryHandler);
```
Comment (shumkov) [inline lib/createDIContainer.js:415]: Let's just keep`/identities`

File: lib/createDIContainer.js [hunk 8]
```diff
@@ -411,25 +405,19 @@ async function createDIContainer(options) {
...
+      router.on('GET', '/dataContracts/by-id', dataContractQueryHandler);
```
Comment (shumkov) [inline lib/createDIContainer.js:416]: Let's just keep `/dataContracts`

File: lib/createDIContainer.js [hunk 9]
```diff
@@ -411,25 +405,19 @@ async function createDIContainer(options) {
...
+      router.on('GET', '/dataContracts/documents/by-type', documentQueryHandler);
```
Comment (shumkov) [inline lib/createDIContainer.js:417]: Let's just keep `/dataContracts/documents`

File: lib/document/mongoDbRepository/getDocumentDatabaseFactory.js [hunk 10]
```diff
@@ -7,12 +7,12 @@
...
   /**
    * @typedef getDocumentDatabase
-   * @param {string} dataContractId
+   * @param {Buffer} dataContractId
    * @return {Db}
    */
   async function getDocumentDatabase(dataContractId) {
     const documentMongoDBClient = await connectToDocumentMongoDB();
-    return documentMongoDBClient.db(`${documentMongoDBPrefix}${dataContractId}`);
+    return documentMongoDBClient.db(`${documentMongoDBPrefix}${dataContractId.toString()}`);
```
Comment (shumkov) [inline lib/document/mongoDbRepository/getDocumentDatabaseFactory.js:15]: We need to make sure that `dataContractId` is an `Identifier` (`new Identifier(dataContractId).toString()`)

**Review by shumkov — DISMISSED** [review]
Good job, guys! 👍

**Review by antouhou — APPROVED** [review]

**Pull request merged**